### PR TITLE
Incorrect utf8toUnicode transformation for 00xx

### PR DESCRIPTION
**Pull request by marcstern**
Fix issue and restructure handling
https://github.com/owasp-modsecurity/ModSecurity/issues/3129

Perform all the checks and calculate the multibute length, then write the bytes at a single place (the correct way).

### CODE REVIEW

File: apache2/msc_util.c [hunk 1]
```diff
@@ -105,225 +105,101 @@ int swap_int32(int x) {
...
     int unicode_len = 0, length = 0;
-    unsigned int d = 0, count = 0;
+    unsigned int d = 0;
     unsigned char c, *utf;
     char *rval, *data;
     unsigned int i, len, j;
     unsigned int bytes_left = input_len;
     unsigned char *unicode = NULL;
 
+    if (input == NULL) return NULL;
```
Comment (airween) [inline apache2/msc_util.c]: Is `NULL` allowed here? If not, we could use `assert()`.

**Comment by airween** [comment]
Hi @marcstern,

I added one comment above, and would like to ask if we modify a function that used by a transformation (`utf8toUnicode`) then probably we should add a test case to our CI. What do you think?

File: apache2/msc_util.c (hunk 1 above)
Comment (marcstern) [reply apache2/msc_util.c]: I changed it

**Review by airween — APPROVED** [review]

**Comment by airween** [comment]
@marcstern do you think we can add a test case to check this transformation's behavior? Or it's not necessary?

**Comment by marcstern** [comment]
We should ideally add  test for this, indeed

**Pull request merged**